Multiple bus container registration

phatboyg opened the pull request:
## Changes in Progress

- [x] Move to `IMyBus` and `MyBus` types, which are used to proxy `IBus`, but named by the developer and used as the dependency instead of having generic `IBus<T>` interfaces. Since this interface is defined by the developer, they can add their own properties to it that have meaning to them, making it actually useful while the core proxy calls are handled by the base proxy class (see next).
- [x] Use a base `BusInstance<T>` proxy class as a generic type constraint, to force developers onto the right path.

Maybe: 

-  ~~If `IMyBus` includes `IMyPublishEndpoint`, and `IMyPublishEndpoint` is based upon `IPublishEndpoint` - that interface will be registered to resolve to the proper bus instance.~~
-  ~~If `IMyBus` includes `IMySendEndpointProvider`, and `IMySendEndpointProvider` is based upon `ISendEndpointProvider` - that interface will be registered to resolve to the proper bus instance.~~

> The general guidance would be to just use the `IMyBus` as a dependency to send messages to the other bus instance, instead of breaking it out into the various components.



## Things To Investigate

- [x] ASP.NET Core Health Checks should include all bus instances
  - [x] This should be abstracted into the core assembly as a way to ask the bus for its health without any dependencies, and then map that to the health check components in ASP.NET when the publisher requests it. This would address the common question of "how do I know my bus is connected."
- [ ] Coordination of Outbox across bus instances, since you know it will be expected.

NooNameR commented:
what do you think about such scenario:
```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddMassTransit<IMyBus, MyBus>(cfg =>
    {
        cfg.UseEndpointProvider<IMyEndpointProvider, MyEndpointProvider>();
        cfg.UsePublishEndpoint<>();
    });
}
interface IMyEndpointProvider
{
}
class MyEndpointProvider : IMyEndpointProvider
{
    public MyEndpointProvider(ISendEndpointProvider)
    {
    }
}
interface IMyBus
{
}
class MyBus : IMyBus
{
    public MyBus(IBus bus)
    {
    }
}
```

it could be also used without any interfaces just classes

phatboyg commented:
Having slept on it, I don’t think supporting `IPublishEndpoint` or `ISendEndpointProvider` specialization is necessary, nor does it add any value. If anything, it creates confusion and leads people to do _stupid stuff_.

Components used by a consumer should not be aware of the outer infrastructure components, such as `IMyBus`. I’m inclined to say that stuff should stay on the outer edge of the service. If they want to publish events from the domain objects to a separate bus instance, to me that just seems wrong.

And if they really want to publish to `IMyBus`, they can using it. They don’t need `IMyPublishEndpoint` to do it.

ntziolis commented:
I didn't yet have time to review the PR in detail (will do so of the we), but from my experiencing trying to add this on top specialization of the existing lib ([document here for reference](https://stackoverflow.com/a/61589404/268742)):
- I'd agree that we can do without `SendEndpoint` (assuming same is true for PublishEndpoint) specialization as we can generate one by leveraging the specialised `IBus`
- If you know you need a specific instance and are not relying on the context, I think its fair to go through the `IBus` 
  - To be fair I didn't fully grasp the difference between `IBus.getSendEndpoint()` and injecting `ISendEndpoint` diretly but I'm awar that doc calls out to use `IBus` as last resort
  - But as long as you @phatboyg give the green light to to access the send endpoint via the bus in this situation the approach of not specializing Send/Publish endpoint should be just fine
- I did run into an issue that I wasn't able to solve without injecting a non generic interface and that was request/response
  - This is due to the fact that the client is generated by a factory inside the library itself
  - To get this to work I made sure to provide a factory method which is not spezialized but ALWAYS returns a client factory from a spezialized hardcoded `IBus`
```c#
// TODO check where this factory is used, likely in consumer context.RespondAsync
services.AddSingleton(provider => provider.GetRequiredService<IConfigurationServiceProvider>().GetRequiredService<IBus<IHosted>>().Instance.CreateClientFactory());
```

Again I didn't have time to fully understand where the client instance gets generated, but I wanted to call out that this was the one issue I wasn't able to solve in a spezialized manner. It be great if we could make sure that whatever approach we are taking also covers request/response situations.

ntziolis commented:
After a quick look through the code I was delighted to see you are already going the direction of a generic `ServiceCollectionConfigurator`, which was exactly what I was missing to get request/response working in specialized manner.

phatboyg commented:
@alexeyzimarev I removed the obsolete (simplified) hosted service bits, and adjusted the health checks to be located within the bus (since multiple bus instances are now supported in this branch). If you could take a quick peek and see if I missed anything, that would be great. The unit test for health checks passes.

NooNameR commented:
@phatboyg health check updates :)
https://github.com/NooNameR/MassTransit/commit/9c0ef6bfe015c4ed8e6b76003222d7744d811043

alexeyzimarev commented:
Thanks for the heads-up, will check in the morning

On Fri, 8 May 2020 at 21:24, Chris Patterson <notifications@github.com>
wrote:

> @alexeyzimarev <https://github.com/alexeyzimarev> I removed the obsolete
> (simplified) hosted service bits, and adjusted the health checks to be
> located within the bus (since multiple bus instances are now supported in
> this branch). If you could take a quick peek and see if I missed anything,
> that would be great. The unit test for health checks passes.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MassTransit/MassTransit/pull/1864#issuecomment-625977536>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVQYVOAAVATEHJOFOFT2PLRQRL7PANCNFSM4M3XSSWA>
> .
>

phatboyg commented:
Documentation added: https://masstransit-project.com/usage/containers/multibus.html

(once it builds anyway).
